### PR TITLE
[Catalog Promotions] Reapplying Catalog Promotion on Action Edit

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        Given the store operates on a channel identified by "Web-US" code
+        And the store has a "T-Shirt" configurable product
+        And this product has "PHP T-Shirt" variant priced at "$20.00" in "Web-US" channel
+        And there is a catalog promotion "Winter sale" that reduces price by "50%" and applies on "PHP T-shirt" variant
+        And I am logged in as an administrator
+
+    @api
+    Scenario: Reapplying catalog promotion after editing action
+        Given I want to modify a catalog promotion "Winter sale"
+        When I edit its action so that it reduces price by "25%"
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:17]:
```suggestion
        When I modify a catalog promotion "Winter sale"
        And I edit its action so that it reduces price by "25%"
```

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        And I save my changes
+        Then I log out
+        When I view "PHP T-Shirt" variant
+        Then the product variant price should be "$15.00"
+        And the product original price should be "$20.00"
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:22]: 😱 
```suggestion
        And the visitor views "PHP T-Shirt" variant
        Then the product variant price should be "$15.00"
        And the product original price should be "$20.00"
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 3]
```diff
@@ -164,6 +164,7 @@ public function iWantToModifyACatalogPromotion(CatalogPromotionInterface $catalo
 
     /**
      * @When /^I add action that gives ("[^"]+") percentage discount$/
+     * @When /^I edit its action so that it reduces price by ("[^"]+")$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:167]: Are you sure, that this step edits the existing action and not adds the new one?

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        And there is a catalog promotion "Winter sale" that reduces price by "50%" and applies on "PHP T-shirt" variant
+        And I am logged in as an administrator
+
+    @api
+    Scenario: Reapplying catalog promotion after editing action
+        When I modify a catalog promotion "Winter sale"
+        And I edit its action so that it reduces price by "25%"
+        And I save my changes
+        Then the visitor view "PHP T-Shirt" variant
+        Then the product variant price should be "$15.00"
```
Comment (AdamKasp) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:20]:
```suggestion
        Then this product variant price should be "$15.00"
```
?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml [hunk 5]
```diff
@@ -129,6 +129,8 @@ Example configuration for `percentage_discount` action type:
             </attribute>
         </property>
         <property name="rules" writable="true" />
-        <property name="actions" writable="true" />
+        <property name="actions" writable="true">
+            <subresource resourceClass="%sylius.model.catalog_promotion_action.class%" collection="true" />
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotion.xml:133]: we don't want to have it as a subresource, because actions (and rules) make sense only with catalog promotion.
So IMO we should find some different solution.

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 6]
```diff
@@ -0,0 +1,21 @@
+@applying_catalog_promotions
+Feature: Reapplying catalog promotions after editing its action
```
Comment (Zales0123) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:2]: `catalog promotion` without `s`? Or `its` -> `theirs` 💃

File: src/Sylius/Bundle/CoreBundle/EventListener/CatalogPromotionEventListener.php [hunk 7]
```diff
@@ -32,7 +33,7 @@ public function postPersist(LifecycleEventArgs $args): void
     {
         $entity = $args->getObject();
 
-        if ($entity instanceof CatalogPromotionInterface) {
+         if ($entity instanceof CatalogPromotionInterface) {
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/EventListener/CatalogPromotionEventListener.php:36]: 😱

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature (hunk 4 above)
Comment (arti0090) [reply features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:20]: This step is used also in other feature files, should I also change them? Because context is similar.

**Comment by arti0090** [comment]
No path for action 🎉 


image

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature (hunk 4 above)
Comment (AdamKasp) [reply features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:20]: imo not necessary, but if you want to fix it in  more places I suggest to open separate PR

File: src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        $paths = (array) $docs['paths'];
+
+        foreach ($this->apiRoutes as $apiRoute) {
+            if (array_key_exists($apiRoute, $paths)) {
+                unset($paths[$apiRoute]);
+            }
+        }
+
+        $docs['paths'] = $paths;
+        return $docs;
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php:47]: what do you think about sth like:
```suggestion
        $docs = $this->decoratedNormalizer->normalize($object, $format, $context);

        foreach ($this->apiRoutes as $apiRoute) {
            if (array_key_exists($apiRoute, $paths)) {
                unset($docs['paths'][$apiRoute]);
            }
        }

        return $docs;
```

File: src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Swagger;
+
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+class PathHiderDocumentationNormalizer implements NormalizerInterface
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php:18]: you should add spec for this class

File: src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php (hunk 8 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php:47]: Don't treat $docs['paths'] as usuall array because it is ArrayObject 😄 . Without conversion the usuall methods does not work.
Also i tried to remove line 46 ->  $docs['paths'] = $paths;  and without this the swagger has some trouble resolving the frontend documentation. 🤔

File: src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php (hunk 9 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php:18]: We don't have specs for any DocumentationNormalizer in our codebase 🦖

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+    Background:
+        Given the store operates on a channel identified by "Web-US" code
+        And the store has a "T-Shirt" configurable product
+        And this product has "PHP T-Shirt" variant priced at "$20.00" in "Web-US" channel
+        And there is a catalog promotion "Winter sale" that reduces price by "50%" and applies on "PHP T-shirt" variant
+        And I am logged in as an administrator
+
+    @api
+    Scenario: Reapplying catalog promotion after editing action
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:15]:
```suggestion
    Scenario: Reapplying catalog promotion after editing its action
```

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        When I modify a catalog promotion "Winter sale"
+        And I edit its action so that it reduces price by "25%"
+        And I save my changes
+        Then the visitor view "PHP T-Shirt" variant
+        Then the product variant price should be "$15.00"
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:20]:
```suggestion
        And the product variant price should be "$15.00"
```

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotionAction.xml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+                <attribute name="output">false</attribute>
+            </itemOperation>
+        </itemOperations>
+
+        <collectionOperations/>
+
+        <property name="id" identifier="true" writable="false" />
+        <property name="catalogPromotion" readable="true" />
+        <property name="type" readable="true" />
+        <property name="configuration" readable="true" />
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/CatalogPromotionAction.xml:36]: Not needed I guess
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php (hunk 8 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php:47]:
```suggestion

        return $docs;
```

File: src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    public function supportsNormalization($data, $format = null): bool
+    {
+        return $this->decoratedNormalizer->supportsNormalization($data, $format);
+    }
+
+    public function normalize($object, $format = null, array $context = [])
+    {
+        $docs = $this->decoratedNormalizer->normalize($object, $format, $context);
+        $paths = (array) $docs['paths'];
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php:38]:
```suggestion
        $paths = $docs['paths']->getArray();
```
or sth similar?

**Review by GSadee — COMMENTED** [review]
Add more scenarios to cover different behaviours, e.g.:

- removing an existing action
- adding a new action
- removing an existing action and adding a new one

File: src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php (hunk 13 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/Swagger/PathHiderDocumentationNormalizer.php:38]: I can use $paths = $docs['paths']->getArrayCopy();
which is returning array.

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    Scenario: Reapplying catalog promotion after removing and adding its action
+        When I modify a catalog promotion "Winter sale"
+        And I remove its every action
+        And I save my changes
+        Then I modify a catalog promotion "Winter sale" again
+        And I add action that gives "10%" percentage discount
+        And I save my changes
+        Then the visitor view "PHP T-Shirt" variant
+        And the product variant price should be "$18.00"
+        And the product original price should be "$20.00"
```
Comment (lchrusciel) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:33]: There should never be `Then` twice in Behat scenario. I'm also not convinced, we should make saves twice in such scenario
```suggestion
    @api
    Scenario: Reapplying catalog promotion after removing and adding its action
        When I modify a catalog promotion "Winter sale"
        And I remove its every action
        And I add action that gives "10%" percentage discount
        And I save my changes
        Then the visitor view "PHP T-Shirt" variant
        And the product variant price should be "$18.00"
        And the product original price should be "$20.00"
```

Comment (arti0090) [reply features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature]: May I do it in another PR pls?

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_adding_its_action.feature [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    Background:
+        Given the store operates on a channel identified by "Web-US" code
+        And the store has a "T-Shirt" configurable product
+        And this product has "PHP T-Shirt" variant priced at "$20.00" in "Web-US" channel
+        And there is a catalog promotion with "Winter_sale" code and "Winter sale" name
+        And it applies on "PHP T-Shirt" variant
+        And I am logged in as an administrator
+
+    @api
+    Scenario: Reapplying catalog promotion after adding its action
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_adding_its_action.feature:16]:
```suggestion
    Scenario: Reapplying catalog promotion after adding a new action to it
```

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature (hunk 4 above)
Comment (GSadee) [reply features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:20]: But you can just add a new step with `this`

File: features/promotion/managing_catalog_promotions/creating_catalog_promotion.feature [hunk 16]
```diff
@@ -44,4 +42,3 @@ Feature: Creating a catalog promotion
         Then there should be 1 new catalog promotion on the list
         And it should have "winter_sale" code and "Winter sale" name
         And the catalog promotion "Winter sale" should be available in channel "United States"
-        And this catalog promotion should be usable
```
Comment (GSadee) [inline features/promotion/managing_catalog_promotions/creating_catalog_promotion.feature:47]: Why did you remove these steps?

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 17]
```diff
@@ -154,6 +153,7 @@ public function iRenameTheCatalogPromotionTo(CatalogPromotionInterface $catalogP
 
     /**
      * @When I want to modify a catalog promotion :catalogPromotion
+     * @When I modify a catalog promotion :catalogPromotion (again)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:156]:
```suggestion
     * @When I modify a catalog promotion :catalogPromotion
```

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_adding_its_action.feature [hunk 18]
```diff
@@ -0,0 +1,22 @@
+@applying_catalog_promotions
+Feature: Reapplying catalog promotion after editing its action
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_adding_its_action.feature:2]: In the second feature file you also have scenarios with adding actions, I would be for merging these two files into one

File: features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+    @api
+    Scenario: Reapplying catalog promotion after changing the effect of the promotion in action configuration
```
Comment (GSadee) [inline features/promotion/applying_catalog_promotions/reapplying_catalog_promotions_after_editing_its_action.feature:36]:
```suggestion
    Scenario: Reapplying catalog promotion after adding another action
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 20]
```diff
@@ -465,14 +504,6 @@ public function itShouldHaveRule(ProductVariantInterface $firstVariant, ProductV
         );
     }
 
-    /**
-     * @Then this catalog promotion should be usable
-     */
-    public function thisCatalogPromotionShouldBeUsable(): void
-    {
-        Assert::isInstanceOf($this->messageBus->getDispatchedMessages()[0]['message'], CatalogPromotionCreated::class);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:473]: You should probably change the event name here

File: features/promotion/managing_catalog_promotions/creating_catalog_promotion.feature (hunk 16 above)
Comment (arti0090) [reply features/promotion/managing_catalog_promotions/creating_catalog_promotion.feature:47]: I have restored it only for scenario where CP has an action

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :tada:
